src/se/vidstige/jadb/server/AdbProtocolHandler.java

```diff
@@ -37,91 +37,111 @@ public void run() {
-                        writer.writeBytes(d.getSerial() + "\t" + d.getType() + "\n");
-                    }
-                    output.writeBytes("OKAY");
-                    send(output, new String(tmp.toByteArray(), StandardCharsets.UTF_8));
-                } else if (command.startsWith("host:transport:")) {
-                    String serial = command.substring("host:transport:".length());
-                    selected = findDevice(serial);
-                    output.writeBytes("OKAY");
-                } else if ("sync:".equals(command)) {
-                    output.writeBytes("OKAY");
-                    try {
-                        sync(output, input);
-                    } catch (JadbException e) { // sync response with a different type of fail message
-                        SyncTransport sync = new SyncTransport(output, input);
-                        sync.send("FAIL", e.getMessage());
-                    }
-                } else if (command.startsWith("shell:")) {
-                    String shellCommand = command.substring("shell:".length());
-                    output.writeBytes("OKAY");
-                    shell(shellCommand, output, input);
-                    output.close();
-                    return;
-                } else if ("host:get-state".equals(command)) {
-                    // TODO: Check so that exactly one device is selected.
-                    AdbDeviceResponder device = responder.getDevices().get(0);
-                    output.writeBytes("OKAY");
-                    send(output, device.getType());
-                } else if (command.startsWith("host-serial:")) {
-                    String[] strs = command.split(":",0);
-                    if (strs.length != 3) {
-                        throw new ProtocolException("Invalid command: " + command);
-                    }
-
-                    String serial = strs[1];
-                    boolean found = false;
-                    output.writeBytes("OKAY");
-                    for (AdbDeviceResponder d : responder.getDevices()) {
-                        if (d.getSerial().equals(serial)) {
-                            send(output, d.getType());
-                            found = true;
-                            break;
-                        }
-                    }
-
-                    if (!found) {
-                        send(output, "unknown");
-                    }
-                } else {
-                    throw new ProtocolException("Unknown command: " + command);
-                }
-            } catch (ProtocolException e) {
-                output.writeBytes("FAIL");
-                send(output, e.getMessage());
+        try (
+            DataInputStream input = new DataInputStream(socket.getInputStream());
+            DataOutputStream output = new DataOutputStream(socket.getOutputStream())
+        ) {
+            while (processCommand(input, output)) {
+                // nothing to do here
```

Comment:
This seems weird. `processCommand` has side effects. It would be better to do

```java
while (true) {
    processCommand(input, output)
}
```

Reply:
processCommand needs to break the cycle.
Therefore:
```java
while (true) {
    if (!processCommand(...)) {
        break;
   }
}
```